File: scripts/memory_decay.py
#!/usr/bin/env python3
import json
import os
import re
import sys
import uuid
from math import sqrt
from pathlib import Path
from datetime import datetime, timezone

STORE = Path(os.getcwd()) / 'store'
VALID_TYPES = {'decision', 'experiment', 'reference', 'status', 'temporary'}
VALID_TTLS = {'3d', '7d', '30d', 'permanent'}
TIER_WEIGHT = {'fresh': 1.0, 'recent': 0.85, 'faded': 0.6, 'ghost': 0.3, 'expired': 0}


def ensure_dirs(root=STORE):
    for d in ('fresh', 'archive', 'expired'):
        (root / d).mkdir(parents=True, exist_ok=True)
    idx = root / 'index.json'
    if not idx.exists():
        idx.write_text('[]\n', encoding='utf8')
    return root


def read_index(root=STORE):
    idx = root / 'index.json'
    if not idx.exists():
        return []
    return json.loads(idx.read_text(encoding='utf8'))


def write_index(entries, root=STORE):
    (root / 'index.json').write_text(json.dumps(entries, indent=2, ensure_ascii=False) + '\n', encoding='utf8')


def storage_dir_for_tier(tier):
    if tier == 'expired':
        return 'expired'
    if tier in ('ghost', 'faded'):
        return 'archive'
    return 'fresh'


def write_memory(entry, body, root=STORE):
    ensure_dirs(root)
    path = root / storage_dir_for_tier(entry['tier']) / f"{entry['id']}.json"
    path.write_text(json.dumps({**entry, 'body': body}, indent=2, ensure_ascii=False) + '\n', encoding='utf8')
    index = read_index(root)
    for i, item in enumerate(index):
        if item['id'] == entry['id']:
            index[i] = dict(entry)
            break
    else:
        index.append(dict(entry))
    write_index(index, root)


def read_memory(mem_id, root=STORE):
    for d in ('fresh', 'archive', 'expired'):
        path = root / d / f'{mem_id}.json'
        if path.exists():
            return json.loads(path.read_text(encoding='utf8'))
    return None


def move_memory(mem_id, old_dir, new_dir, root=STORE):
    src = root / old_dir / f'{mem_id}.json'
    dst = root / new_dir / f'{mem_id}.json'
    if src.exists() and src != dst:
        dst.parent.mkdir(parents=True, exist_ok=True)
        src.rename(dst)


def count_files(d, root=STORE):
    p = root / d
    return len(list(p.glob('*.json'))) if p.exists() else 0


def parse_ttl(ttl):
    if ttl == 'permanent':
        return float('inf')
    m = re.match(r'^(\d+)d$', ttl or '')
    return int(m.group(1)) * 86400000 if m else 30 * 86400000


def age_days(created):
    dt = datetime.fromisoformat(created.replace('Z', '+00:00'))
    return (datetime.now(timezone.utc) - dt).total_seconds() / 86400


def tier_for_age(days):
    if days <= 3:
        return 'fresh'
    if days <= 14:
        return 'recent'
    if days <= 30:
        return 'faded'
    return 'ghost'


def compute_decay(entry):
    if entry.get('ttl') == 'permanent':
        return entry.get('tier', 'fresh')
    days = age_days(entry['created'])
    ttl_days = parse_ttl(entry.get('ttl', '30d')) / 86400000
    if days > ttl_days:
        return 'expired'
    return tier_for_age(days)


def run_decay(dry_run=False):
    index = read_index(STORE)
    changes = []
    for entry in index:
        nxt = compute_decay(entry)
        if nxt != entry.get('tier'):
            changes.append({'id': entry['id'], 'summary': entry['summary'], 'domain': entry['domain'], 'from': entry.get('tier'), 'to': nxt})
            if not dry_run:
                old_dir = storage_dir_for_tier(entry.get('tier'))
                new_dir = storage_dir_for_tier(nxt)
                if old_dir != new_dir:
                    move_memory(entry['id'], old_dir, new_dir, STORE)
                entry['tier'] = nxt
    if not dry_run:
        write_index(index, STORE)
    return changes


def tokenize(text):
    if not text:
        return []
    tokens = []
    tokens.extend(re.findall(r'[a-z0-9_\-.]+', text.lower()))
    for seg in re.findall(r'[\u4e00-\u9fff]+', text):
        for i in range(len(seg) - 1):
            tokens.append(seg[i:i+2])
        tokens.extend(seg)
    return tokens


def term_freq(tokens):
    tf = {}
    for t in tokens:
        tf[t] = tf.get(t, 0) + 1
    return tf


def cosine_sim(a, b):
    dot = sum(a[t] * b.get(t, 0) for t in a)
    mag_a = sum(v * v for v in a.values())
    mag_b = sum(v * v for v in b.values())
    return 0 if not mag_a or not mag_b else dot / (sqrt(mag_a) * sqrt(mag_b))


def load_domain_aliases():
    cfg = STORE / 'config.json'
    if not cfg.exists():
        return {}
    try:
        return json.loads(cfg.read_text(encoding='utf8')).get('domainAliases', {})
    except Exception:
        return {}


def domain_alias_boost(query, domain):
    aliases = load_domain_aliases().get(domain)
    if not aliases:
        return 0
    q = query.lower()
    boost = sum(0.4 for alias in aliases if str(alias).lower() in q)
    return min(boost, 0.8)


def score_entry(q_tokens, entry, query):
    summary_score = cosine_sim(q_tokens, term_freq(tokenize(entry.get('summary', ''))))
    domain_score = cosine_sim(q_tokens, term_freq(tokenize(entry.get('domain', '')))) * 3
    alias_boost = domain_alias_boost(query, entry.get('domain'))
    tier_w = TIER_WEIGHT.get(entry.get('tier'), 0.5)
    return (summary_score + domain_score + alias_boost) * tier_w


def display_summary(summary, tier):
    return f"[archived] {summary[:15]}..." if tier == 'ghost' else summary


def search(query, limit=5):
    q = term_freq(tokenize(query))
    scored = []
    for entry in read_index(STORE):
        if entry.get('tier') == 'expired':
            continue
        score = score_entry(q, entry, query)
        if score > 0:
            item = dict(entry)
            item['score'] = score
            scored.append(item)
    scored.sort(key=lambda x: x['score'], reverse=True)
    out = []
    for e in scored[:limit]:
        if e.get('tier') in ('fresh', 'recent'):
            full = read_memory(e['id'], STORE)
            if full:
                e['body'] = full.get('body')
        out.append(e)
    return out


def scan(query):
    q = term_freq(tokenize(query))
    domains = {}
    for entry in read_index(STORE):
        if entry.get('tier') == 'expired':
            continue
        d = entry.get('domain', 'uncategorized')
        domains.setdefault(d, {'domain': d, 'count': 0, 'latest': None, 'maxScore': 0})
        domains[d]['count'] += 1
        score = score_entry(q, entry, query)
        domains[d]['maxScore'] = max(domains[d]['maxScore'], score)
        if not domains[d]['latest'] or entry['created'] > domains[d]['latest']:
            domains[d]['latest'] = entry['created']
    return sorted([d for d in domains.values() if d['maxScore'] > 0], key=lambda x: x['maxScore'], reverse=True)


def focus(domain):
    items = [e for e in read_index(STORE) if e.get('domain') == domain and e.get('tier') != 'expired']
    items.sort(key=lambda x: x['created'], reverse=True)
    out = []
    for e in items:
        item = dict(e)
        if item.get('tier') in ('fresh', 'recent'):
            full = read_memory(item['id'], STORE)
            if full:
                item['body'] = full.get('body')
        out.append(item)
    return out


def parse_args(argv):
    opts = {'_': []}
    i = 0
    while i < len(argv):
        a = argv[i]
        if a.startswith('--'):
            key = a[2:]
            if i + 1 >= len(argv) or argv[i + 1].startswith('-'):
                opts[key] = True
            else:
                opts[key] = argv[i + 1]
                i += 1
        elif a.startswith('-'):
            key = a[1:]
            if i + 1 >= len(argv) or argv[i + 1].startswith('-'):
                opts[key] = True
            else:
                opts[key] = argv[i + 1]
                i += 1
        else:
            opts['_'].append(a)
        i += 1
    return opts


def fail(msg):
    print(msg, file=sys.stderr)
    sys.exit(1)


def validate_write(opts):
    mem_type = opts.get('type') or opts.get('t')
    domain = opts.get('domain') or opts.get('d')
    summary = opts.get('summary') or opts.get('s')
    ttl = opts.get('ttl', '30d')
    body = opts.get('body') or opts.get('b') or summary
    try:
        confidence = float(opts.get('confidence') or opts.get('c') or '0.8')
    except ValueError:
        fail('Confidence must be between 0.0 and 1.0.')
    if mem_type not in VALID_TYPES:
        fail(f'Invalid type: {mem_type}')
    if not domain or not re.match(r'^[A-Za-z0-9_-]+$', domain):
        fail('Domain is required and should be a simple tag.')
    if not summary or len(summary.strip()) < 5:
        fail('Summary is required.')
    if ttl not in VALID_TTLS:
        fail(f'Invalid ttl: {ttl}')
    if not (0 <= confidence <= 1):
        fail('Confidence must be between 0.0 and 1.0.')
    return mem_type, domain, summary.strip(), ttl, confidence, body


def print_help():
    print('memory_decay.py commands:\n  write --type decision --domain infra --summary "..." [--ttl 30d] [--confidence 0.8] [--body "..."]\n  search <query> [--limit 5]\n  scan <query>\n  focus <domain>\n  decay [--dry-run]\n  stats')


def main():
    ensure_dirs(STORE)
    argv = sys.argv[1:]
    if not argv or argv[0] in ('help', '--help', '-h'):
        print_help()
        return
    cmd, rest = argv[0], argv[1:]
    if cmd == 'write':
        mem_type, domain, summary, ttl, confidence, body = validate_write(parse_args(rest))
        entry = {
            'id': str(uuid.uuid4()),
            'created': datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z'),
            'type': mem_type,
            'domain': domain,
            'summary': summary,
            'ttl': ttl,
            'confidence': confidence,
            'tier': 'fresh',
        }
        write_memory(entry, body, STORE)
        print(f"Written: [{entry['type']}] {entry['summary']}")
        print(f"id: {entry['id']} | domain: {entry['domain']} | ttl: {entry['ttl']} | tier: fresh")
    elif cmd == 'search':
        args = parse_args(rest)
        query = args['_'][0] if args['_'] else None
        if not query:
            fail('Query required.')
        for e in search(query, int(args.get('limit') or args.get('n') or 5)):
            print(f"[{e['tier']}] {e['created'][:10]} | {e['domain']} | {display_summary(e['summary'], e['tier'])} | {e['score'] * 100:.0f}%")
            if e.get('body') and e.get('tier') in ('fresh', 'recent'):
                body = str(e['body'])
                print('  ' + body[:200] + ('...' if len(body) > 200 else ''))
    elif cmd == 'scan':
        args = parse_args(rest)
        query = args['_'][0] if args['_'] else None
        if not query:
            fail('Query required.')
        for d in scan(query):
            print(f"{d['domain']} ({d['count']} memories, latest: {(d['latest'] or '?')[:10]}, relevance: {d['maxScore'] * 100:.0f}%)")
    elif cmd == 'focus':
        args = parse_args(rest)
        domain = args['_'][0] if args['_'] else None
        if not domain:
            fail('Domain required.')
        for e in focus(domain):
            print(f"[{e['tier']}] {e['created'][:10]}: {e['summary']}")
            if e.get('body'):
                body = str(e['body'])
                print('  ' + body[:200] + ('...' if len(body) > 200 else ''))
    elif cmd == 'decay':
        args = parse_args(rest)
        for c in run_decay(bool(args.get('dry-run'))):
            print(f"{c['summary']} [{c['domain']}]: {c['from']} -> {c['to']}")
    elif cmd == 'stats':
        index = read_index(STORE)
        tiers, domains, types = {}, {}, {}
        for e in index:
            tiers[e['tier']] = tiers.get(e['tier'], 0) + 1
            domains[e['domain']] = domains.get(e['domain'], 0) + 1
            types[e['type']] = types.get(e['type'], 0) + 1
        print(f'Total: {len(index)}')
        print(f"Files: fresh={count_files('fresh')}, archive={count_files('archive')}, expired={count_files('expired')}")
        print('By tier:')
        for k, v in tiers.items():
            print(f'  {k}: {v}')
        print('By domain:')
        for k, v in sorted(domains.items(), key=lambda kv: kv[1], reverse=True):
            print(f'  {k}: {v}')
        print('By type:')
        for k, v in types.items():
            print(f'  {k}: {v}')
    else:
        fail(f'Unknown command: {cmd}')


if __name__ == '__main__':
    main()

File: scripts/memory_decay.js
#!/usr/bin/env node
import { mkdirSync, existsSync, readFileSync, writeFileSync, renameSync, readdirSync } from 'fs';
import { join } from 'path';
import { randomUUID } from 'crypto';

const STORE = join(process.cwd(), 'store');
const VALID_TYPES = new Set(['decision', 'experiment', 'reference', 'status', 'temporary']);
const VALID_TTLS = new Set(['3d', '7d', '30d', 'permanent']);
const TIER_WEIGHT = { fresh: 1.0, recent: 0.85, faded: 0.6, ghost: 0.3, expired: 0 };

function ensureDirs(root = STORE) {
  for (const d of ['fresh', 'archive', 'expired']) mkdirSync(join(root, d), { recursive: true });
  const idx = join(root, 'index.json');
  if (!existsSync(idx)) writeFileSync(idx, '[]\n', 'utf8');
  return root;
}

function readIndex(root = STORE) {
  const idx = join(root, 'index.json');
  if (!existsSync(idx)) return [];
  return JSON.parse(readFileSync(idx, 'utf8'));
}

function writeIndex(entries, root = STORE) {
  writeFileSync(join(root, 'index.json'), JSON.stringify(entries, null, 2) + '\n', 'utf8');
}

function storageDirForTier(tier) {
  return tier === 'expired' ? 'expired' : (tier === 'ghost' || tier === 'faded') ? 'archive' : 'fresh';
}

function writeMemory(entry, body, root = STORE) {
  ensureDirs(root);
  const file = join(root, storageDirForTier(entry.tier), `${entry.id}.json`);
  writeFileSync(file, JSON.stringify({ ...entry, body }, null, 2) + '\n', 'utf8');
  const index = readIndex(root);
  const i = index.findIndex((e) => e.id === entry.id);
  if (i >= 0) index[i] = { ...entry };
  else index.push({ ...entry });
  writeIndex(index, root);
}

function readMemory(id, root = STORE) {
  for (const dir of ['fresh', 'archive', 'expired']) {
    const file = join(root, dir, `${id}.json`);
    if (existsSync(file)) return JSON.parse(readFileSync(file, 'utf8'));
  }
  return null;
}

function moveMemory(id, fromDir, toDir, root = STORE) {
  const src = join(root, fromDir, `${id}.json`);
  const dst = join(root, toDir, `${id}.json`);
  if (existsSync(src) && src !== dst) renameSync(src, dst);
}

function countFiles(dir, root = STORE) {
  const p = join(root, dir);
  if (!existsSync(p)) return 0;
  return readdirSync(p).filter((f) => f.endsWith('.json')).length;
}

function parseTTL(ttl) {
  if (ttl === 'permanent') return Infinity;
  const m = String(ttl || '').match(/^(\d+)d$/);
  return m ? parseInt(m[1], 10) * 86400000 : 30 * 86400000;
}

function ageDays(created) {
  return (Date.now() - new Date(created).getTime()) / 86400000;
}

function tierForAge(days) {
  if (days <= 3) return 'fresh';
  if (days <= 14) return 'recent';
  if (days <= 30) return 'faded';
  return 'ghost';
}

function computeDecay(entry) {
  if (entry.ttl === 'permanent') return entry.tier || 'fresh';
  const days = ageDays(entry.created);
  const ttlDays = parseTTL(entry.ttl || '30d') / 86400000;
  if (days > ttlDays) return 'expired';
  return tierForAge(days);
}

function runDecay(dryRun = false) {
  const index = readIndex(STORE);
  const changes = [];
  for (const entry of index) {
    const next = computeDecay(entry);
    if (next !== entry.tier) {
      changes.push({ id: entry.id, summary: entry.summary, domain: entry.domain, from: entry.tier, to: next });
      if (!dryRun) {
        const oldDir = storageDirForTier(entry.tier);
        const newDir = storageDirForTier(next);
        if (oldDir !== newDir) moveMemory(entry.id, oldDir, newDir, STORE);
        entry.tier = next;
      }
    }
  }
  if (!dryRun) writeIndex(index, STORE);
  return changes;
}

function tokenize(text) {
  if (!text) return [];
  const tokens = [];
  const eng = text.toLowerCase().match(/[a-z0-9_\-.]+/g);
  if (eng) tokens.push(...eng);
  const cjk = text.match(/[\u4e00-\u9fff]+/g);
  if (cjk) {
    for (const seg of cjk) {
      for (let i = 0; i < seg.length - 1; i++) tokens.push(seg.slice(i, i + 2));
      for (const ch of seg) tokens.push(ch);
    }
  }
  return tokens;
}

function termFreq(tokens) {
  const tf = {};
  for (const t of tokens) tf[t] = (tf[t] || 0) + 1;
  return tf;
}

function cosineSim(a, b) {
  let dot = 0;
  let magA = 0;
  let magB = 0;
  for (const t in a) {
    magA += a[t] ** 2;
    if (b[t]) dot += a[t] * b[t];
  }
  for (const t in b) magB += b[t] ** 2;
  if (!magA || !magB) return 0;
  return dot / (Math.sqrt(magA) * Math.sqrt(magB));
}

function loadDomainAliases() {
  const configPath = join(STORE, 'config.json');
  if (!existsSync(configPath)) return {};
  try {
    return JSON.parse(readFileSync(configPath, 'utf8')).domainAliases || {};
  } catch {
    return {};
  }
}

function domainAliasBoost(query, domain) {
  const aliases = loadDomainAliases()[domain];
  if (!aliases) return 0;
  const q = query.toLowerCase();
  let boost = 0;
  for (const alias of aliases) {
    if (q.includes(String(alias).toLowerCase())) boost += 0.4;
  }
  return Math.min(boost, 0.8);
}

function scoreEntry(qTokens, entry, query) {
  const summaryScore = cosineSim(qTokens, termFreq(tokenize(entry.summary || '')));
  const domainScore = cosineSim(qTokens, termFreq(tokenize(entry.domain || ''))) * 3;
  const aliasBoost = domainAliasBoost(query, entry.domain);
  const tierW = TIER_WEIGHT[entry.tier] || 0.5;
  return (summaryScore + domainScore + aliasBoost) * tierW;
}

function displaySummary(summary, tier) {
  if (tier === 'ghost') return `[archived] ${summary.slice(0, 15)}...`;
  return summary;
}

function search(query, limit = 5) {
  const index = readIndex(STORE);
  const qTokens = termFreq(tokenize(query));
  const scored = [];
  for (const entry of index) {
    if (entry.tier === 'expired') continue;
    const score = scoreEntry(qTokens, entry, query);
    if (score > 0) scored.push({ ...entry, score });
  }
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, limit).map((e) => {
    const result = { ...e };
    if (e.tier === 'fresh' || e.tier === 'recent') {
      const full = readMemory(e.id, STORE);
      if (full) result.body = full.body;
    }
    return result;
  });
}

function scan(query) {
  const index = readIndex(STORE);
  const qTokens = termFreq(tokenize(query));
  const domains = {};
  for (const entry of index) {
    if (entry.tier === 'expired') continue;
    const domain = entry.domain || 'uncategorized';
    if (!domains[domain]) domains[domain] = { domain, count: 0, latest: null, maxScore: 0 };
    domains[domain].count++;
    const score = scoreEntry(qTokens, entry, query);
    if (score > domains[domain].maxScore) domains[domain].maxScore = score;
    if (!domains[domain].latest || new Date(entry.created) > new Date(domains[domain].latest)) {
      domains[domain].latest = entry.created;
    }
  }
  return Object.values(domains).filter((d) => d.maxScore > 0).sort((a, b) => b.maxScore - a.maxScore);
}

function focus(domain) {
  return readIndex(STORE)
    .filter((e) => e.domain === domain && e.tier !== 'expired')
    .sort((a, b) => new Date(b.created) - new Date(a.created))
    .map((e) => {
      const result = { ...e };
      if (e.tier === 'fresh' || e.tier === 'recent') {
        const full = readMemory(e.id, STORE);
        if (full) result.body = full.body;
      }
      return result;
    });
}

function parseArgs(argv) {
  const out = { _: [] };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a.startsWith('--')) {
      const key = a.slice(2);
      const next = argv[i + 1];
      if (!next || next.startsWith('-')) out[key] = true;
      else {
        out[key] = next;
        i++;
      }
    } else if (a.startsWith('-')) {
      const key = a.slice(1);
      const next = argv[i + 1];
      if (!next || next.startsWith('-')) out[key] = true;
      else {
        out[key] = next;
        i++;
      }
    } else {
      out._.push(a);
    }
  }
  return out;
}

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

function validateWrite(opts) {
  const type = opts.type || opts.t;
  const domain = opts.domain || opts.d;
  const summary = opts.summary || opts.s;
  const ttl = opts.ttl || '30d';
  const confidence = parseFloat(opts.confidence || opts.c || '0.8');
  if (!VALID_TYPES.has(type)) fail(`Invalid type: ${type}`);
  if (!domain || !/^[a-z0-9_-]+$/i.test(domain)) fail('Domain is required and should be a simple tag.');
  if (!summary || summary.trim().length < 5) fail('Summary is required.');
  if (!VALID_TTLS.has(ttl)) fail(`Invalid ttl: ${ttl}`);
  if (Number.isNaN(confidence) || confidence < 0 || confidence > 1) fail('Confidence must be between 0.0 and 1.0.');
  return { type, domain, summary: summary.trim(), ttl, confidence, body: opts.body || opts.b || summary };
}

function printHelp() {
  console.log(`memory_decay.js commands:\n  write --type decision --domain infra --summary "..." [--ttl 30d] [--confidence 0.8] [--body "..."]\n  search <query> [--limit 5]\n  scan <query>\n  focus <domain>\n  decay [--dry-run]\n  stats`);
}

ensureDirs(STORE);
const [cmd, ...rest] = process.argv.slice(2);

if (!cmd || cmd === 'help' || cmd === '--help' || cmd === '-h') {
  printHelp();
} else if (cmd === 'write') {
  const opts = validateWrite(parseArgs(rest));
  const entry = {
    id: randomUUID(),
    created: new Date().toISOString(),
    type: opts.type,
    domain: opts.domain,
    summary: opts.summary,
    ttl: opts.ttl,
    confidence: opts.confidence,
    tier: 'fresh',
  };
  writeMemory(entry, opts.body, STORE);
  console.log(`Written: [${entry.type}] ${entry.summary}`);
  console.log(`id: ${entry.id} | domain: ${entry.domain} | ttl: ${entry.ttl} | tier: fresh`);
} else if (cmd === 'search') {
  const args = parseArgs(rest);
  const query = args._[0];
  if (!query) fail('Query required.');
  const limit = parseInt(args.limit || args.n || '5', 10);
  const results = search(query, limit);
  if (!results.length) console.log('No matches.');
  for (const e of results) {
    console.log(`[${e.tier}] ${e.created.slice(0, 10)} | ${e.domain} | ${displaySummary(e.summary, e.tier)} | ${(e.score * 100).toFixed(0)}%`);
    if (e.body && (e.tier === 'fresh' || e.tier === 'recent')) {
      console.log(`  ${String(e.body).slice(0, 200)}${String(e.body).length > 200 ? '...' : ''}`);
    }
  }
} else if (cmd === 'scan') {
  const args = parseArgs(rest);
  const query = args._[0];
  if (!query) fail('Query required.');
  const results = scan(query);
  if (!results.length) console.log('No matching domains found.');
  for (const d of results) {
    console.log(`${d.domain} (${d.count} memories, latest: ${(d.latest || '?').slice(0, 10)}, relevance: ${(d.maxScore * 100).toFixed(0)}%)`);
  }
} else if (cmd === 'focus') {
  const args = parseArgs(rest);
  const domain = args._[0];
  if (!domain) fail('Domain required.');
  const results = focus(domain);
  if (!results.length) console.log(`No memories in domain "${domain}".`);
  for (const e of results) {
    console.log(`[${e.tier}] ${e.created.slice(0, 10)}: ${e.summary}`);
    if (e.body) console.log(`  ${String(e.body).slice(0, 200)}${String(e.body).length > 200 ? '...' : ''}`);
  }
} else if (cmd === 'decay') {
  const args = parseArgs(rest);
  const changes = runDecay(Boolean(args['dry-run']));
  if (!changes.length) console.log('No decay needed.');
  for (const c of changes) {
    console.log(`${c.summary} [${c.domain}]: ${c.from} -> ${c.to}`);
  }
} else if (cmd === 'stats') {
  const index = readIndex(STORE);
  const tiers = {};
  const domains = {};
  const types = {};
  for (const e of index) {
    tiers[e.tier] = (tiers[e.tier] || 0) + 1;
    domains[e.domain] = (domains[e.domain] || 0) + 1;
    types[e.type] = (types[e.type] || 0) + 1;
  }
  console.log(`Total: ${index.length}`);
  console.log(`Files: fresh=${countFiles('fresh')}, archive=${countFiles('archive')}, expired=${countFiles('expired')}`);
  console.log('By tier:');
  for (const [k, v] of Object.entries(tiers)) console.log(`  ${k}: ${v}`);
  console.log('By domain:');
  for (const [k, v] of Object.entries(domains).sort((a, b) => b[1] - a[1])) console.log(`  ${k}: ${v}`);
  console.log('By type:');
  for (const [k, v] of Object.entries(types)) console.log(`  ${k}: ${v}`);
} else {
  fail(`Unknown command: ${cmd}`);
}
